pip_audit/_dependency_source/requirement.py

```diff
@@ -164,16 +138,38 @@ def _fix_file(self, filename: Path, fix_version: ResolvedFixVersion) -> None:
 
         # Now write out the new requirements file
         with filename.open("w") as f:
+            fixed = False
             for req in req_list:
                 if (
                     req.name == fix_version.dep.name
                     and req.specifier.contains(fix_version.dep.version)
                     and not req.specifier.contains(fix_version.version)
                 ):
                     req.specifier = SpecifierSet(f"=={fix_version.version}")
+                    fixed = True
                 assert req.marker is None or req.marker.evaluate()
                 f.write(str(req) + os.linesep)
 
+            # The vulnerable dependency may not be explicitly listed in the requirements file if it
+            # is a subdependency of a requirement. In this case, we should explicitly add the fixed
+            # dependency into the requirements file.
+            #
+            # To know whether this is the case, we'll need to resolve dependencies if we haven't
+            # already in order to figure out whether this subdependency belongs to this file or
+            # another.
+            try:
+                if not fixed and fix_version.dep in self._collect_cached_deps(
+                    filename, list(reqs.values())
+                ):
+                    logger.warning(
+                        f"added fixed subdependency explicitly to requirements file {filename}: "
+                        f"{fix_version.dep.canonical_name}"
+                    )
+                    f.write("# added by pip-audit to fix subdependency" + os.linesep)
+                    f.write(f"{fix_version.dep.canonical_name}=={fix_version.version}" + os.linesep)
```

Comment:
If I'm reading this right, this will produce outputs like this, right?

```
# added by pip-audit to fix subdependency
some-subdep==1.2.3
```

IMO we should do something closer to what `pip-compile` does, for example:

```
some-subdep==1.2.3
    # pip-audit: subdependency fixed via some-dep (4.5.6)
```

Note the indentation of the comment, and the context we provide. OTOH, I'm not sure how much of a pain it'll be to retrieve that context.

We should do the same when a fixed subdep is beneath multiple deps, e.g.:

```
some-subdep==1.2.3
    # pip-audit: subdependency fixed via some-dep (4.5.6), other-dep (7.8.9)
```

Reply:
Good idea. As discussed, I'll do this.